bblfsh/pyuast.c

```diff
@@ -14,77 +14,231 @@ static const char *ReadStr(const void *data, const char *prop)
   PyObject *node = (PyObject *)data;
   PyObject *children_obj = PyObject_GetAttrString(node, "children");
-  if (!children_obj)
+  if (!children_obj) {
     return NULL;
+  }
 
   PyObject *seq = PySequence_Fast(children_obj, "expected a sequence");
   return PyList_GET_ITEM(seq, index);
 }
 
-static int RolesSize(const void *node)
-{
+static int RolesSize(const void *node) {
   return ReadLen(node, "roles");
 }
 
-static uint16_t RoleAt(const void *data, int index)
-{
+static uint16_t RoleAt(const void *data, int index) {
   PyObject *node = (PyObject *)data;
   PyObject *roles_obj = PyObject_GetAttrString(node, "roles");
-  if (!roles_obj)
+  if (!roles_obj) {
     return 0;
+  }
+
   PyObject *seq = PySequence_Fast(roles_obj, "expected a sequence");
   return (uint16_t)PyLong_AsUnsignedLong(PyList_GET_ITEM(seq, index));
 }
 
-static int PropertiesSize(const void *data)
-{
+static int PropertiesSize(const void *data) {
   PyObject *node = (PyObject *)data;
   PyObject *properties_obj = PyObject_GetAttrString(node, "properties");
-  if (!properties_obj)
+  if (!properties_obj) {
     return 0;
+  }
 
   return (int)PyLong_AsLong(properties_obj);
 }
 
-static const char *PropertyAT(const void *data, int index)
-{
+static const char *PropertyKeyAt(const void *data, int index) {
   PyObject *node = (PyObject *)data;
-  PyObject *properties_obj = PyObject_GetAttrString(node, "properties");
-  if (!properties_obj)
+  PyObject *properties = PyObject_GetAttrString(node, "properties");
+  if (!properties) {
+    return NULL;
+  }
+  if (!PyMapping_Check(properties)) {
+    return NULL;
+  }
+  PyObject *keys = PyMapping_Keys(properties);
+  if (!keys) {
+    return NULL;
+  }
+
+  PyObject *seq = PySequence_Fast(keys, "expected a sequence");
+  return PyUnicode_AsUTF8(PyList_GET_ITEM(seq, index));
+}
+
+static const char *PropertyValueAt(const void *data, int index) {
+  PyObject *node = (PyObject *)data;
+  PyObject *properties = PyObject_GetAttrString(node, "properties");
+  if (!properties) {
+    return NULL;
+  }
+  if (!PyMapping_Check(properties)) {
     return NULL;
+  }
+  PyObject *values = PyMapping_Values(properties);
+  if (!values) {
+    return NULL;
+  }
 
-  PyObject *seq = PySequence_Fast(properties_obj, "expected a sequence");
+  PyObject *seq = PySequence_Fast(values, "expected a sequence");
   return PyUnicode_AsUTF8(PyList_GET_ITEM(seq, index));
 }
 
+static bool HasStartOffset(const void *data) {
+  PyObject *node = (PyObject *)data;
+  if (!node) {
+    return false;
+  }
+  return PyObject_GetAttrString(node, "start_position") != Py_None;
+}
+
+static uint32_t StartOffset(const void *data) {
+  PyObject *node = (PyObject *)data;
+  PyObject *position = PyObject_GetAttrString(node, "start_position");
+  if (!position || position == Py_None) {
+    return 0;
+  }
+
+  PyObject *offset = PyObject_GetAttrString(position, "offset");
+  if (!offset) {
+    return 0;
+  }
+  return PyLong_AsUnsignedLong(offset);
```

Comment:
I think a cast would be more explicit here since `PyLong_AsUnsignedLong` returns an `unsigned long`:

`return (uint32_t)PyLong_AsUnsignedLong(offset);`

Same for the other `uint32_t` functions.

Reply:
I didn't do it since I thought function signature was explicit enough in this case. I can add it though.